src/abstractdatatable/join.jl

```diff
@@ -3,217 +3,170 @@
+    DataTableJoiner{DT1,DT2}(dtl, dtr, on)
 
-    ## (left_sorter, left_indexer, leftonly_indexer,
-    ##  right_sorter, right_indexer, rightonly_indexer)
-    (left_sorter[left_indexer], left_sorter[leftonly_indexer],
-     right_sorter[right_indexer], right_sorter[rightonly_indexer])
+# helper map between the row indices in original and joined table
+immutable RowIndexMap
+    "row indices in the original table"
+    orig::Vector{Int}
+    "row indices in the resulting joined table"
+    join::Vector{Int}
 end
 
-function sharepools{S,N}(v1::Union{CategoricalArray{S,N}, NullableCategoricalArray{S,N}},
-                         v2::Union{CategoricalArray{S,N}, NullableCategoricalArray{S,N}},
-                         index::Vector{S},
-                         R)
-    tidx1 = convert(Vector{R}, indexin(CategoricalArrays.index(v1.pool), index))
-    tidx2 = convert(Vector{R}, indexin(CategoricalArrays.index(v2.pool), index))
-    refs1 = zeros(R, length(v1))
-    refs2 = zeros(R, length(v2))
-    for i in 1:length(refs1)
-        if v1.refs[i] != 0
-            refs1[i] = tidx1[v1.refs[i]]
-        end
+Base.length(x::RowIndexMap) = length(x.orig)
+
+# composes the joined data table using the maps between the left and right
+# table rows and the indices of rows in the result
+function compose_joined_table(joiner::DataTableJoiner,
+                              left_ixs::RowIndexMap, leftonly_ixs::RowIndexMap,
+                              right_ixs::RowIndexMap, rightonly_ixs::RowIndexMap)
+    @assert length(left_ixs) == length(right_ixs)
+    # compose left half of the result taking all left columns
+    all_orig_left_ixs = vcat(left_ixs.orig, leftonly_ixs.orig)
+    if length(leftonly_ixs) > 0
+        # permute the indices to restore left table rows order
+        all_orig_left_ixs[vcat(left_ixs.join, leftonly_ixs.join)] = all_orig_left_ixs
+    end
+    ril = length(right_ixs)
+    loil = length(leftonly_ixs)
+    roil = length(rightonly_ixs)
+    left_dt = DataTable(Any[resize!(col[all_orig_left_ixs], length(all_orig_left_ixs)+roil)
+                            for col in columns(joiner.dtl)],
+                        names(joiner.dtl))
+
+    # compose right half of the result taking all right columns excluding on
+    dtr_noon = without(joiner.dtr, joiner.on_cols)
+    # permutation to swap rightonly and leftonly rows
+    right_perm = vcat(1:ril, ril+roil+1:ril+roil+loil, ril+1:ril+roil)
+    if length(leftonly_ixs) > 0
+        # compose right_perm with the permutation that restores left rows order
+        right_perm[vcat(right_ixs.join, leftonly_ixs.join)] = right_perm[1:ril+loil]
     end
-    for i in 1:length(refs2)
-        if v2.refs[i] != 0
-            refs2[i] = tidx2[v2.refs[i]]
+    all_orig_right_ixs = vcat(right_ixs.orig, rightonly_ixs.orig)
+    right_dt = DataTable(Any[resize!(col[all_orig_right_ixs], length(all_orig_right_ixs)+loil)[right_perm]
```

Comment:
I agree that it could be much more efficient, and probably `right_perm` weirdness is not needed at all. Originally it was a workaround for slow `setindex!(PooledDataArray, ix, v)`. As was suggested before, let's make sure the PR passes the tests and merge it. I will try to optimize this method in a separate PR. Unless, of course, there are some general concerns about the efficiency of the suggested approach that have to be addressed before merging.